cleanrl/rnd_ppo.py

```diff
@@ -0,0 +1,1079 @@
+        )
+        self.extra_layer = nn.Sequential(
+            layer_init(nn.Linear(448, 448), std=0.1),
+            nn.ReLU()
+        )
+        self.actor = nn.Sequential(
+            layer_init(nn.Linear(448, 448), std=0.01), 
+            nn.ReLU(),  
+            layer_init(nn.Linear(448, envs.action_space.n), std=0.01))
+        self.critic_ext = layer_init(nn.Linear(448, 1), std=0.01)
+        self.critic_int = layer_init(nn.Linear(448, 1), std=0.01)
+
+    def forward(self, x):
+        return self.network(x)
+
+    def get_action(self, x, action=None):
+        logits = self.actor(self.forward(x))
+        probs = Categorical(logits=logits)
+        if action is None:
+            action = probs.sample()
+        return action, probs.log_prob(action), probs.entropy()
+
+    def get_value(self, x):
+        features = self.forward(x)
+        return self.critic_ext(self.extra_layer(features) + features), self.critic_int(self.extra_layer(features) + features)
+
+agent = Agent().to(device)
+
+rnd_model = RNDModel(4, envs.action_space.n).to(device)
+
+optimizer = optim.Adam(list(agent.parameters()) + list(rnd_model.predictor.parameters()), lr=args.learning_rate, eps=1e-5)
+if args.anneal_lr:
+    # https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/defaults.py#L20
+    lr = lambda f: f * args.learning_rate
+
+reward_rms = RunningMeanStd()
+obs_rms = RunningMeanStd(shape=(1, 1, 84, 84))
+
+discounted_reward = RewardForwardFilter(args.int_gamma)
+
+
+# ALGO Logic: Storage for epoch data
+obs = torch.zeros((args.num_steps, args.num_envs) + envs.observation_space.shape).to(device)
+actions = torch.zeros((args.num_steps, args.num_envs)).to(device)
+logprobs = torch.zeros((args.num_steps, args.num_envs)).to(device)
+rewards = torch.zeros((args.num_steps, args.num_envs)).to(device)
+curiosity_rewards = torch.zeros((args.num_steps, args.num_envs)).to(device)
+dones = torch.zeros((args.num_steps, args.num_envs)).to(device)
+ext_values = torch.zeros((args.num_steps, args.num_envs)).to(device)
+int_values = torch.zeros((args.num_steps, args.num_envs)).to(device)
+
+# TRY NOT TO MODIFY: start the game
+global_step = 0
+# Note how `next_obs` and `next_done` are used; their usage is equivalent to
+# https://github.com/ikostrikov/pytorch-a2c-ppo-acktr-gail/blob/84a7582477fb0d5c82ad6d850fe476829dddd2e1/a2c_ppo_acktr/storage.py#L60
+next_obs = envs.reset()
+next_done = torch.zeros(args.num_envs).to(device)
+num_updates = args.total_timesteps // args.batch_size
+
+print('Start to initailize observation normalization parameter.....')
+next_ob = []
+for step in range(args.num_steps * 50):
+    acs = torch.from_numpy(np.random.randint(0, envs.action_space.n, size=(args.num_envs,)))
+    s, r, d, infos = envs.step(acs)
+    next_ob = s[:, 3, :, :].reshape([-1, 1, 84, 84])
+
+    if len(next_ob) % (args.num_steps * args.num_envs) == 0:
+        obs_rms.update(next_ob)
+        next_ob = []
+del acs
```

Comment:
I assume this will be taken care for by the garbage collector. I suggest removing it.